docs: add colab column (#583)
jina-ai/finetuner: docs/notebooks/text_to_text.md

---
jupyter:
  jupytext:
    text_representation:
      extension: .md
      format_name: markdown
      format_version: '1.3'
      jupytext_version: 1.14.1
  kernelspec:
    display_name: Python 3
    name: python3
---

<!-- #region id="Huf1E2zq7JWb" -->
# Text-to-Text Search via BERT

<a href="https://colab.research.google.com/drive/1Ui3Gw3ZL785I7AuzlHv3I0-jTvFFxJ4_?usp=sharing"><img alt="Open In Colab" src="https://colab.research.google.com/assets/colab-badge.svg"></a>

> Searching large amounts of text documents with text queries is a very popular use-case, so of course Finetuner enables you to accomplish this easily.

Reviewer: ```suggestion
Searching large amounts of text documents with text queries is a very popular use-case and Finetuner enables you to accomplish this easily.
```

In this example, we will fine-tune `bert-base-cased` on the [Quora Question Pairs](https://www.sbert.net/examples/training/quora_duplicate_questions/README.html?highlight=quora#dataset) dataset, where the search task involves finding duplicate questions in the dataset. An example query for this search task might look as follows:

```
How can I be a good geologist?

```

Retrieved documents that could be duplicates for this question should then be ranked in the following order:

```
What should I do to be a great geologist?
How do I become a geologist?
What do geologists do?
...

```

We can fine-tune BERT so that questions that are duplicates of each other are represented in close proximity and questions that are not duplicates will have representations that are further apart in the embedding space. In this way, we can rank the embeddings in our search space by their proximity to the query question and return the highest ranking duplicates.
<!-- #endregion -->

<!-- #region id="SfR6g0E_8fOz" -->
## Data

> We will use the [Quora Question Pairs](https://www.sbert.net/examples/training/quora_duplicate_questions/README.html?highlight=quora#dataset) dataset to show-case Finetuner for text to text search. We have already pre-processed this dataset and made it available for you to pull from hubble. Do this as follows:

Reviewer: ```suggestion
We will use the [Quora Question Pairs](https://www.sbert.net/examples/training/quora_duplicate_questions/README.html?highlight=quora#dataset) dataset to show-case Finetuner for text to text search. We have already pre-processed this dataset and made it available for you to pull from Jina AI Cloud. Do this as follows:
```

<!-- #endregion -->

```python id="pwS11Nsg7jPM"
import finetuner
from docarray import DocumentArray, Document

finetuner.notebook_login()
```

```python id="8PIO5T--p4tR"
train_data = DocumentArray.pull('quora_train.da', show_progress=True)
query_data = DocumentArray.pull('quora_query_dev.da', show_progress=True)
index_data = DocumentArray.pull('quora_index_dev.da', show_progress=True)

train_data.summary()
```

<!-- #region id="r_IlEIp59g9v" -->
So we have 104598 training `Document`s. Each `Document` consists of a text field that contains the question, as well as a `finetuner_label` which indicates the label to which the question belongs. If multiple questions have the same label, they are duplicates of one another. If they have different `finetuner_label`s, they are not duplicates of each other.

> As for the evaluation dataset, we load `query_data` and `index_data` separately. The `query_data` has the same structure as the `train_data`, consisting of labelled documents. The `index_data` is the data against which the queries will be matched, and contains many documents, some of which may be irrelevant to the queries (ie. they have no duplicated in the `query_data`).

Reviewer: ```suggestion
As for the evaluation dataset, we load `query_data` and `index_data` separately. The `query_data` have the same structure as the `train_data`, consisting of labeled documents. The `index_data` are the data against which the queries will be matched, and contain many documents, some of which may be irrelevant to the queries (i.e. they have no duplicates in the `query_data`).
```

> If you look at the summaries for the `query_data` and `index_data`, you will find that they have the following number of instances:

Reviewer: ```suggestion
If you look at the summaries for the `query_data` and `index_data`, you will find that they have the following number of samples:
```

This will make it easy for us to retrieve the experiment and run in the future. We also provide a short description of our run, just for some extra context. 

For this run, we select Finetuner's `TripletMarginLoss` and `TripletMarginMiner`, as they are most relevant for our use-case. The `TripletMarginLoss` measures the similarity between three tensors, namely the anchor, a positive sample and a negative sample. This makes sense for our task, since we want duplicate questions to have representations closer together, while non-duplicates should have more dissimilar representations. Likewise, the `TripletMarginMiner` outputs a tuple of size 3, with an anchor, a positive sample and a negative sample.

Lastly, we provide an `EvaluationCallback` with our `query_data` and `index_data`. This evaluation is done at the end of each epoch and its results will be visible to us in the logs, which we will monitor in the next section. Since we have not specified which metrics should be applied, default metrics will be computed. The `Evaluation` section of this guide will show you the default metrics.
<!-- #endregion -->

<!-- #region id="h0DGNRo8-lZD" -->
## Monitoring

> Now that we've created a run, let's see its status. You can monitor the run by checking the status - `run.status()`, the logs - `run.logs()` or `run.stream_logs()`. 

Reviewer: ```suggestion
Now that we've created a run, let's see its status. You can monitor the run by checking the status - `run.status()` - and the logs - `run.logs()` or `run.stream_logs()`. 
```

*note, the job will take around 15 minutes to finish.*
<!-- #endregion -->

```python id="gajka0TG-S6u"
for entry in run.stream_logs():
    print(entry)
```

<!-- #region id="7AuB0IWC_CSt" -->

> Dependends on the size of the training data, some runs might take up to several hours, you can reconnect to your run very easily to monitor its status.

Reviewer: ```suggestion
Depending on the size of the training data, some runs might take up to several hours. You can later reconnect to your run easily to monitor its status.
```
